### PR TITLE
Bug/mov 723 multiple exceptions on logout

### PR DESCRIPTION
**Pull request by hb0**

### CODE REVIEW

File: synchronization/src/main/java/de/cyface/synchronization/SyncAdapter.java [hunk 1]
```diff
@@ -124,6 +125,12 @@ public void onPerformSync(@NonNull final Account account, @NonNull final Bundle
         // This allows us to mock the #isConnected() check for unit tests
         mockIsConnectedToReturnTrue = extras.containsKey(MOCK_IS_CONNECTED_TO_RETURN_TRUE);
 
+        // If WifiSurveyor.shutdownSurveillance was called in the meantime, cancel sync:
```
Comment (hb0) [inline synchronization/src/main/java/de/cyface/synchronization/SyncAdapter.java]: We only want to have the hard ValidationException below if the sync was not interrupted by us before (e.g. logout), which is a the only legitimate case where the token may be null.

File: synchronization/src/main/java/de/cyface/synchronization/WiFiSurveyor.java [hunk 2]
```diff
@@ -233,6 +233,10 @@ public void stopSurveillance() throws SynchronisationException {
                 throw new SynchronisationException(e);
             }
         }
+
+        // This interrupts ongoing synchronization or else it may crash because required data is missing (token,
+        // account)
+        ContentResolver.cancelSync(currentSynchronizationAccount, authority);
```
Comment (hb0) [inline synchronization/src/main/java/de/cyface/synchronization/WiFiSurveyor.java]: sets the interrupted() flag to true

File: synchronization/src/main/java/de/cyface/synchronization/WiFiSurveyor.java [hunk 3]
```diff
@@ -226,6 +226,7 @@ public void stopSurveillance() throws SynchronisationException {
                 return;
             }
             connectivityManager.unregisterNetworkCallback(networkCallback);
+            networkCallback = null;
```
Comment (hb0) [inline synchronization/src/main/java/de/cyface/synchronization/WiFiSurveyor.java]: forgot this before. This prevents a crash when you log out (which unregisters the network callbacks) and kill the app, which also unregister the callback if it's not null.

File: datacapturing/src/cyface/java/de/cyface/datacapturing/CyfaceDataCapturingService.java [hunk 4]
```diff
@@ -154,6 +155,25 @@ public void startWifiSurveyor() throws SetupException {
         }
     }
 
+    /**
```
Comment (hb0) [inline datacapturing/src/cyface/java/de/cyface/datacapturing/CyfaceDataCapturingService.java]: that's the same code as MovebisDataCapturingService.deregisterJwtToken executes, which was what I wanted to test in our app to reproduce and fix bugs. That's why I added the logout functionality to our app, which calls this method here.

File: synchronization/src/test/java/de/cyface/synchronization/WiFiSurveyorTest.java [hunk 5]
```diff
@@ -271,7 +271,14 @@ private void setConnectivity(final boolean connected, final boolean wifiNotMobil
         // Now we can call the NetworkCallbacks with the correct networkCapabilities (registration is ok)
         // Only call the networkCallback for wifi connections as we only register those in production code
         if (wifiNotMobile || !surveyor.isSyncOnUnMeteredNetworkOnly()) {
-            Validate.isTrue(shadowConnectivityManager.getNetworkCallbacks().size() == 1);
+            final boolean isAccountRegistered = oocut.currentSynchronizationAccount != null;
+            assertThat(shadowConnectivityManager.getNetworkCallbacks().size(),
+                    is(equalTo(isAccountRegistered ? 1 : 0)));
+
+            if (!isAccountRegistered) {
+                return; // Without accounts, not network callbacks must be updates
```
Comment (muthenberg) [inline synchronization/src/test/java/de/cyface/synchronization/WiFiSurveyorTest.java]: type

**Pull request merged**